scripts/roll_release.sh

```diff
@@ -0,0 +1,119 @@
+#!/usr/bin/env bash
+
+#
+# Distributed Linear Algebra with Future (DLAF)
+#
+# Copyright (c) 2018-2024, ETH Zurich
+# All rights reserved.
+#
+# Please, refer to the LICENSE file in the root directory.
+# SPDX-License-Identifier: BSD-3-Clause
+#
+
+# This script tags a release locally and creates a release on GitHub. It relies
+# on the hub command line tool (https://hub.github.com/).
+
+set -o errexit
+
+VERSION_MAJOR=$(sed -n 's/project(DLAF VERSION \([0-9]\+\)\.[0-9]\+\.[0-9]\+)/\1/p' CMakeLists.txt)
+VERSION_MINOR=$(sed -n 's/project(DLAF VERSION [0-9]\+\.\([0-9]\+\)\.[0-9]\+)/\1/p' CMakeLists.txt)
+VERSION_PATCH=$(sed -n 's/project(DLAF VERSION [0-9]\+\.[0-9]\+\.\([0-9]\+\))/\1/p' CMakeLists.txt)
+VERSION_FULL="${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}"
+VERSION_FULL_TAG="v${VERSION_MAJOR}.${VERSION_MINOR}.${VERSION_PATCH}"
+VERSION_TITLE="DLA-Future ${VERSION_FULL}"
+CURRENT_BRANCH="$(git rev-parse --abbrev-ref HEAD)"
+RELEASE_BRANCH="release-${VERSION_MAJOR}.${VERSION_MINOR}"
+
+if ! which hub >/dev/null 2>&1; then
+    echo "Hub not installed on this system (see https://hub.github.com/). Exiting."
+    exit 1
+fi
+
+if ! [[ "$CURRENT_BRANCH" == "$RELEASE_BRANCH" ]]; then
+    echo "Not on release branch (expected \"$RELEASE_BRANCH\", currently on \"${CURRENT_BRANCH}\"). Not continuing to make release."
+    exit 1
+fi
+
+changelog_path="CHANGELOG.md"
+readme_path="README.md"
+
+echo "You are about to tag and create a final release on GitHub."
+
+echo ""
+echo "Sanity checking release"
+
+sanity_errors=0
+
+printf "Checking that %s has an entry for %s... " "${changelog_path}" "${VERSION_FULL}"
+if grep "## DLA-Future ${VERSION_FULL}" "${changelog_path}"; then
+    echo "OK"
+else
+    echo "Missing"
+    sanity_errors=$((sanity_errors + 1))
+fi
+
+printf "Checking that %s has a documentation entry for %s... " "${readme_path}" "${VERSION_FULL}"
+if grep "^- \[Documentation of \`${VERSION_FULL_TAG}\`\](https://eth-cscs.github.io/DLA-Future/${VERSION_FULL_TAG}" "${readme_path}"; then
+    echo "OK"
+else
+    echo "Missing"
+    sanity_errors=$((sanity_errors + 1))
+fi
+exit 0
+
+if [[ ${sanity_errors} -gt 0 ]]; then
+    echo "Found ${sanity_errors} error(s). Fix it/them and try again."
+    exit 1
+fi
+
+# Extract the changelog for this version
+VERSION_DESCRIPTION=$(
+    # Find the correct heading and print everything from there to the end of the file
+    awk "/^## DLA-Future ${VERSION_FULL}/,EOF" ${changelog_path} |
+        # Remove the heading
+        tail -n+3 |
+        # Find the next heading or the end of the file and print everything until that heading
+        sed '/^## /Q'
```

Comment:
In this way we would end up with `###` titles in the release note, while all the previous releases had `##`.

Reply:
Yep. We can:
- Update the past changelogs (I'm happy to do that) leaving the format here as it is.
- Update the format of `CHANGELOG.md` to maybe remove the `# Changelog` heading and have everything else moved up one level.
- Replace `^##` by `#` in the release script to get consistent heading levels.

I'm indifferent about these options. Do you have a preference?

Reply:
No strong preference.

Reply:
I did the third option in https://github.com/eth-cscs/DLA-Future/pull/1127/commits/4c10cb59497dd3a50dd177926da81b47bba3084a. This keeps existing released changelogs, and the new `CHANGELOG.md` file unchanged, with a very small change in the release script.